packages/convex-helpers/server/triggers.ts

```diff
@@ -0,0 +1,267 @@
+ * triggers.register("myTableName", async (ctx, change) => {
+ *   console.log("Table changed", change);
+ * });
+ * // Use `mutation` to define all mutations, and the triggers will get called.
+ * export const mutation = customMutation(rawMutation, triggers.customFunctionWrapper());
+ * ```
+ */
+export class Triggers<
+  DataModel extends GenericDataModel,
+  Ctx extends { db: GenericDatabaseWriter<DataModel> } = GenericMutationCtx<DataModel>,
+> {
+  registered: {
+    [TableName in TableNamesInDataModel<DataModel>]?: Trigger<Ctx, DataModel, TableName>[];
+  } = {};
+  
+  register<TableName extends TableNamesInDataModel<DataModel>>(
+    tableName: TableName,
+    trigger: Trigger<Ctx, DataModel, TableName>,
+  ) {
+    if (!this.registered[tableName]) {
+      this.registered[tableName] = [];
+    }
+    this.registered[tableName]!.push(trigger);
+  }
+
+  customFunctionWrapper(): Mod<Ctx, {}, {innerDb: GenericDatabaseWriter<DataModel>}, {}> {
+    const triggers = this;
+    return {
+      args: {},
+      input(ctx: Ctx) {
+        const innerDb = ctx.db;
+        return { ctx: {
+          db: new DatabaseWriterWithTriggers(ctx, innerDb, triggers),
+          innerDb,
+        }, args: {} };
+      },
+    };
+  }
+}
+
+class Lock {
+  promise: Promise<void> | null = null;
+  resolve: (() => void) | null = null;
+
+  async lock(): Promise<(() => void)> {
+    while (this.promise !== null) {
+      await this.promise;
+    }
+    [this.promise, this.resolve] = this._newLock();
+    return () => {
+      this.promise = null;
+      this.resolve?.();
+    };
+  }
+  _newLock(): [Promise<void>, () => void] {
+    let resolve: () => void;
+    const promise = new Promise<void>((r) => {
+      resolve = r;
+    });
+    return [promise, () => resolve()];
+  }
+}
+
+/**
+ * Locking semantics:
+ * - Database writes to tables with triggers are serialized with
+ *   `innerWriteLock` so we can calculate the `change` object without
+ *   interference from parallel writes.
+ * - When the application (not a trigger) calls `insert`, `patch`, or `replace`,
+ *   it will acquire the outer write lock and hold it while doing the write
+ *   operation and all subsequent triggers, including recursive triggers.
+ *   - This ensures atomicity in the simple case where a trigger doesn't call
+ *     other triggers recursively.
+ * - Recursive triggers are queued up, so they are executed in the same order
+ *   as the database writes were. At a high level, this is a BFS traversal of
+ *   the trigger graph.
+ * - Note when there are multiple triggers, they can't be executed atomically
+ *   with the writes that caused them, from the perspective of the other
+ *   triggers. So if one trigger is making sure denormalized data is
+ *   consistent, another trigger could see the data in an inconsistent state.
+ *   To avoid such problems, triggers should be resilient to such
+ *   inconsistencies or the trigger graph should be kept simple.
```

Comment:
do you think it'd be reasonable to just _ban_ recursive triggers for now? that way the semantics would be very straightforward. but perhaps that's too restrictive?

Reply:
you actually get pretty similar guarantees when you have multiple triggers, even if they're not recursive.

like if you have these triggers:

```
triggers.register("users", async (ctx, change) => {
  const isTopStudent = await userAggregate.offsetUntil(ctx, change.newDoc.gpa) < 5;
  if (change.newDoc.isTopStudent !== isTopStudent) {
    ctx.db.patch(change.id, { isTopStudent });
  }
});
triggers.register("users", userAggregate.triggers());
```

one of the triggers has to account for the other not having run yet.

having multiple triggers seems like an important use-case, so if we have those then we might as well have recursive triggers too?

I think the semantics of recursive triggers makes sense, and I wouldn't look forward to banning them -- would it throw an error or just not execute them? -- and explaining those semantics.